### PR TITLE
Allow per-column-family processing in wal_filter (publish log numbers for column family to wal_filter, and provide log number in the record callback)

### PR DESCRIPTION
**Pull request by PraveenSinghRao**
This change is to make wal_filter understand if a given record is applicable to a column_family (since the log record could have been flushed for some column family but not the other, so it may or may not apply to a given column family).

This change contains the following:
1) Column-family name/number map exposed to wal_filter - this is because this is during open, client doesn't have CF handles yet, and can only understand name.
2) CF number/lognumber map exposed to wal_filter - so that wal_filter can know if log is applicable to a given column-family.
3) Providing lognumber with the record to wal_filter callback - so it can check if the given record applies to a column-family.

### CODE REVIEW

**Comment by PraveenSinghRao** [comment]
Not sure the Java related failures in Travis build are related to my change.

File: include/rocksdb/wal_filter.h [hunk 1]
```diff
@@ -30,6 +32,20 @@ class WalFilter {
...
   virtual ~WalFilter() {}
 
+  // Provide ColumnFamily->LogNumber map to filter
+  // so that filter can determine whether given record
+  // is applicable to given column family.
+  // We also pass in name->id map as this is known at
+  // recovery time and write batch callbacks happen
+  // in terms of column family id.
+  //
+  // @params cf_lognumber_map column_family_id to lognumber map
```
Comment (siying) [inline include/rocksdb/wal_filter.h]: Need more description here. What this mapping mean. I guess it means the earliest log number where this column family has data not flushed to SST files?

File: db/db_test.cc [hunk 2]
```diff
@@ -10620,6 +10620,173 @@ TEST_F(DBTest, WalFilterTestWithChangeBatchExtraKeys) {
 
   ValidateKeyExistence(db_, keys_must_exist, keys_must_not_exist);
 }
+
+TEST_F(DBTest, WalFilterTestWithColumnFamilies) {
```
Comment (siying) [inline db/db_test.cc]: Now db_test.cc is too large. We added a db_test2.cc. New tests should go there.

File: db/db_test.cc [hunk 3]
```diff
@@ -10620,6 +10620,173 @@ TEST_F(DBTest, WalFilterTestWithChangeBatchExtraKeys) {
...
+          virtual Status PutCF(uint32_t column_family_id, const Slice& key, 
+            const Slice& /*value*/) override {
+            auto it = cf_log_number_map_.find(column_family_id);
+            assert(it != cf_log_number_map_.end());
+            unsigned long long log_number_for_cf = it->second;
+            // If the current record is applicable for column_family_id
+            // (i.e. isn't flushed to SST file(s) for column_family_id)
+            // add it to the cf_wal_keys_ map for verification.
+            if (log_number_ >= log_number_for_cf) {
+              cf_wal_keys_[column_family_id].push_back(std::string(key.data(), key.size()));
```
Comment (siying) [inline db/db_test.cc]: We are still using 80-char rule.

**Comment by siying** [comment]
Looks good to me.

File: db/db_test.cc (hunk 3 above)
Comment (PraveenSinghRao) [inline db/db_test.cc]: Fixed.

File: db/db_test.cc (hunk 2 above)
Comment (PraveenSinghRao) [inline db/db_test.cc]: Moved all wal_filter related tests to db_test2.cc.

File: include/rocksdb/wal_filter.h (hunk 1 above)
Comment (PraveenSinghRao) [inline include/rocksdb/wal_filter.h]: Yes you are right. Clarified the comment.

File: db/db_test.cc (hunk 2 above)
Comment (yuslepukhin) [inline db/db_test.cc]: @siying Does Linux now run db_test2 as a part of CI and should Windows build do the same?

**Comment by siying** [comment]
Build failure:

16:10:52 In file included from db/db_test2.cc:12:0:
16:10:52 ./include/rocksdb/wal_filter.h: In member function ‘virtual void rocksdb::DBTest2_WalFilterTest_Test::TestBody()’:
16:10:52 ./include/rocksdb/wal_filter.h:78:31: error: ‘virtual rocksdb::WalFilter::WalProcessingOption rocksdb::WalFilter::LogRecord(long long unsigned int, const string&, const rocksdb::WriteBatch&, rocksdb::WriteBatch*, bool*)’ was hidden [-Werror=overloaded-virtual]
16:10:52    virtual WalProcessingOption LogRecord(unsigned long long log_number,
16:10:52                                ^
16:10:52 db/db_test2.cc:127:33: error:   by ‘virtual rocksdb::WalFilter::WalProcessingOption rocksdb::DBTest2_WalFilterTest_Test::TestBody()::TestWalFilter::LogRecord(const rocksdb::WriteBatch&, rocksdb::WriteBatch*, bool*) const’ [-Werror=overloaded-virtual]
16:10:52      virtual WalProcessingOption LogRecord(const WriteBatch& batch,
16:10:52                                  ^
16:10:52 In file included from db/db_test2.cc:12:0:
16:10:52 ./include/rocksdb/wal_filter.h: In member function ‘virtual void rocksdb::DBTest2_WalFilterTestWithChangeBatch_Test::TestBody()’:
16:10:52 ./include/rocksdb/wal_filter.h:78:31: error: ‘virtual rocksdb::WalFilter::WalProcessingOption rocksdb::WalFilter::LogRecord(long long unsigned int, const string&, const rocksdb::WriteBatch&, rocksdb::WriteBatch*, bool*)’ was hidden [-Werror=overloaded-virtual]
16:10:52    virtual WalProcessingOption LogRecord(unsigned long long log_number,
16:10:52                                ^
16:10:52 db/db_test2.cc:323:33: error:   by ‘virtual rocksdb::WalFilter::WalProcessingOption rocksdb::DBTest2_WalFilterTestWithChangeBatch_Test::TestBody()::TestWalFilterWithChangeBatch::LogRecord(const rocksdb::WriteBatch&, rocksdb::WriteBatch*, bool*) const’ [-Werror=overloaded-virtual]
16:10:52      virtual WalProcessingOption LogRecord(const WriteBatch& batch,
16:10:52                                  ^
16:10:52 In file included from db/db_test2.cc:12:0:
16:10:52 ./include/rocksdb/wal_filter.h: In member function ‘virtual void rocksdb::DBTest2_WalFilterTestWithChangeBatchExtraKeys_Test::TestBody()’:
16:10:52 ./include/rocksdb/wal_filter.h:78:31: error: ‘virtual rocksdb::WalFilter::WalProcessingOption rocksdb::WalFilter::LogRecord(long long unsigned int, const string&, const rocksdb::WriteBatch&, rocksdb::WriteBatch*, bool*)’ was hidden [-Werror=overloaded-virtual]
16:10:52    virtual WalProcessingOption LogRecord(unsigned long long log_number,
16:10:52                                ^
16:10:52 db/db_test2.cc:422:33: error:   by ‘virtual rocksdb::WalFilter::WalProcessingOption rocksdb::DBTest2_WalFilterTestWithChangeBatchExtraKeys_Test::TestBody()::TestWalFilterWithChangeBatchAddExtraKeys::LogRecord(const rocksdb::WriteBatch&, rocksdb::WriteBatch*, bool*) const’ [-Werror=overloaded-virtual]
16:10:52      virtual WalProcessingOption LogRecord(const WriteBatch& batch,
16:10:52                                  ^
16:10:52 In file included from db/db_test2.cc:12:0:
16:10:52 ./include/rocksdb/wal_filter.h: In member function ‘virtual void rocksdb::DBTest2_WalFilterTestWithColumnFamilies_Test::TestBody()’:
16:10:52 ./include/rocksdb/wal_filter.h:89:31: error: ‘virtual rocksdb::WalFilter::WalProcessingOption rocksdb::WalFilter::LogRecord(const rocksdb::WriteBatch&, rocksdb::WriteBatch*, bool*) const’ was hidden [-Werror=overloaded-virtual]
16:10:52    virtual WalProcessingOption LogRecord(const WriteBatch& batch,
16:10:52                                ^
16:10:52 db/db_test2.cc:505:33: error:   by ‘virtual rocksdb::WalFilter::WalProcessingOption rocksdb::DBTest2_WalFilterTestWithColumnFamilies_Test::TestBody()::TestWalFilterWithColumnFamilies::LogRecord(long long unsigned int, const string&, const rocksdb::WriteBatch&, rocksdb::WriteBatch*, bool*)’ [-Werror=overloaded-virtual]
16:10:52      virtual WalProcessingOption LogRecord(unsigned long long log_number,
16:10:52                                  ^


Can you fix it?

**Comment by PraveenSinghRao** [comment]
Fixing it. Will PR momentarily.

**Comment by PraveenSinghRao** [comment]
Pull request 1047 for the fix. Thanks and sorry for the break.

**Comment by siying** [comment]
@PraveenSinghRao thanks for your fix! We run different CI internally and in Travis, which causes confusion. We also run more intensive CI test internally. So it's common that we report issues after checking in data.

**Comment by PraveenSinghRao** [comment]
No problem @siying. I imagined a stricter check internally which failed :).

We have lint/valgrind equivalents too internally but haven't had a chance to integrate those with our rocksdb repo. Once we do that, fewer issues should leak out to you.

**Pull request merged**